File: Assets/_Project/Scripts/Platform/LogExport.cs
using UnityEngine;

namespace AdversityRoad.Platform
{
    /// <summary>
    /// 让玩家挑一个**自己能进得去**的目录来放调试日志，并把日志复制过去。
    ///
    /// 【为什么需要它】
    /// Application.persistentDataPath 在安卓上是
    ///   /storage/emulated/0/Android/data/&lt;包名&gt;/files/
    /// Android 11 起 scoped storage 把 Android/data 对文件管理器整个藏了起来，
    /// 玩家拿不到文件；而直接往 /sdcard/Download 写，在 API 29+ 上同样被挡。
    /// 官方留下的口子只有一个：**Storage Access Framework**——玩家用系统选择器
    /// 挑一次目录，应用拿到一份可持久化的写授权，此后往里写不需要任何运行时权限。
    ///
    /// 这个类是 LogExport.java 的 C# 壳，用法与本工程既有的 WebScreen / PipMode 一致。
    /// 非安卓平台（编辑器、桌面）没有这套东西，也不需要——那边直接填路径就行，
    /// 所以这里的方法在非安卓上一律返回"没做成"，由 MoveLogger 走普通路径那条分支。
    /// </summary>
    public static class LogExport
    {
        const string JavaClass = "com.adversityroad.logexport.LogExport";
        /// <summary>接收目录选择结果的 GameObject 名（Java 侧用 UnitySendMessage 回调它）。</summary>
        public const string BridgeObject = "MoveLoggerBridge";

        /// <summary>这个平台支不支持"挑一个目录"。</summary>
        public static bool Supported =>
#if UNITY_ANDROID && !UNITY_EDITOR
            true;
#else
            false;
#endif

        /// <summary>打开系统目录选择器。结果通过 MoveLoggerBridge.OnLogFolderPicked 回来。</summary>
        public static void PickFolder(string chooserTitle)
        {
#if UNITY_ANDROID && !UNITY_EDITOR
            try
            {
                using (var cls = new AndroidJavaClass(JavaClass))
                    cls.CallStatic("pickFolder", BridgeObject, chooserTitle);
            }
            catch (System.Exception e)
            {
                Debug.LogWarning("[LogExport] 打不开目录选择器：" + e.Message);
                var go = GameObject.Find(BridgeObject);
                if (go != null) go.SendMessage("OnLogFolderPicked", "");
            }
#else
            var g = GameObject.Find(BridgeObject);
            if (g != null) g.SendMessage("OnLogFolderPicked", "");
#endif
        }

        /// <summary>把本地文件复制进选定目录。返回文档 URI；失败返回空串。</summary>
        public static string Export(string treeUri, string srcPath, string name, string mime)
        {
#if UNITY_ANDROID && !UNITY_EDITOR
            try
            {
                using (var cls = new AndroidJavaClass(JavaClass))
                    return cls.CallStatic<string>("exportFile", treeUri, srcPath, name, mime);
            }
            catch (System.Exception e)
            {
                Debug.LogWarning("[LogExport] 导出失败：" + e.Message);
                return "";
            }
#else
            return "";
#endif
        }

        /// <summary>选定目录的可读名字（设置面板显示用）。</summary>
        public static string FolderLabel(string treeUri)
        {
#if UNITY_ANDROID && !UNITY_EDITOR
            try
            {
                using (var cls = new AndroidJavaClass(JavaClass))
                    return cls.CallStatic<string>("folderLabel", treeUri);
            }
            catch (System.Exception)
            {
                return treeUri;
            }
#else
            return treeUri;
#endif
        }
    }

    /// <summary>
    /// 目录选择结果的接收端。Java 侧只能按**名字**找 GameObject 回调
    /// （UnitySendMessage 的限制），所以这里要有一个名字固定、常驻不销毁的对象。
    /// </summary>
    public class MoveLoggerBridge : MonoBehaviour
    {
        public static void Ensure()
        {
            if (GameObject.Find(LogExport.BridgeObject) != null) return;
            var go = new GameObject(LogExport.BridgeObject);
            Object.DontDestroyOnLoad(go);
            go.AddComponent<MoveLoggerBridge>();
        }

        // 名字由 Java 侧的 UnitySendMessage 指定，不能改。
        void OnLogFolderPicked(string treeUri)
        {
            if (string.IsNullOrEmpty(treeUri))
            {
                Core.GameEvents.RaiseSubtitle("没有选择目录。");
                return;
            }
            Core.MoveLogger.TreeUri = treeUri;
            Core.MoveLogger.CustomDir = "";   // 选了系统目录就以它为准，免得两个来源打架
            bool ok = Core.MoveLogger.ExportNow();
            Core.GameEvents.RaiseSubtitle(ok
                ? "日志目录已选定：" + LogExport.FolderLabel(treeUri) + "（已导出一份）"
                : "日志目录已选定：" + LogExport.FolderLabel(treeUri));
        }
    }
}

File: Assets/Plugins/Android/com/adversityroad/logexport/LogExport.java
package com.adversityroad.logexport;

import android.app.Activity;
import android.app.Fragment;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.DocumentsContract;

import java.io.FileInputStream;
import java.io.InputStream;
import java.io.OutputStream;

import com.unity3d.player.UnityPlayer;

/**
 * Lets the player choose ANY folder for the debug log, and copies the log there.
 *
 * WHY THIS EXISTS
 * Application.persistentDataPath on Android is /storage/emulated/0/Android/data/<pkg>/files.
 * Since Android 11 that directory is hidden from file managers by scoped storage, so the
 * player literally cannot reach the file. Writing to /sdcard/Download directly is also
 * blocked on API 29+ without legacy storage. The supported way to hand a file to the user
 * is the Storage Access Framework: the user picks a folder once, the app gets a persistable
 * write grant for exactly that folder, and no runtime permission is involved at all.
 *
 * NOTE: this file is deliberately ASCII-only. javac's default encoding on the build machine
 * is not guaranteed to be UTF-8, and a non-ASCII byte in a source file is a hard build
 * failure there. All user facing strings are passed in from C#.
 *
 * Three entry points, all called from C# via AndroidJavaClass.CallStatic:
 *   pickFolder(callbackObject, chooserTitle) - open the system folder picker; the chosen
 *       tree URI arrives at UnitySendMessage(callbackObject, "OnLogFolderPicked", uri).
 *   exportFile(treeUri, srcPath, displayName, mime) - copy a local file into that folder,
 *       overwriting a document of the same name. Returns the document URI, or "" on failure.
 *   folderLabel(treeUri) - a short human readable name for the chosen folder (UI only).
 *
 * Why a headless Fragment (same as GalleryPicker): a Fragment needs no manifest entry,
 * so this plugin never touches AndroidManifest.xml.
 */
public class LogExport extends Fragment {

    private static final int REQUEST_CODE = 45092;
    private static final String METHOD = "OnLogFolderPicked";
    private static String sCallbackObject = "MoveLoggerBridge";
    private static String sChooserTitle = "Choose a folder for the log";

    // ---------------------------------------------------------------- pick

    /** Open the system folder picker. callbackObject = Unity GameObject to answer. */
    public static void pickFolder(String callbackObject, String chooserTitle) {
        if (callbackObject != null && callbackObject.length() > 0) {
            sCallbackObject = callbackObject;
        }
        if (chooserTitle != null && chooserTitle.length() > 0) {
            sChooserTitle = chooserTitle;
        }
        final Activity activity = UnityPlayer.currentActivity;
        if (activity == null || Build.VERSION.SDK_INT < 21) {
            UnityPlayer.UnitySendMessage(sCallbackObject, METHOD, "");
            return;
        }
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    LogExport fragment = new LogExport();
                    activity.getFragmentManager()
                            .beginTransaction()
                            .add(fragment, "AdversityRoadLogExport")
                            .commitAllowingStateLoss();
                } catch (Exception e) {
                    UnityPlayer.UnitySendMessage(sCallbackObject, METHOD, "");
                }
            }
        });
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        try {
            Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION
                    | Intent.FLAG_GRANT_WRITE_URI_PERMISSION
                    | Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION);
            startActivityForResult(Intent.createChooser(intent, sChooserTitle), REQUEST_CODE);
        } catch (Exception e) {
            finishWith("");
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode != REQUEST_CODE) {
            return;
        }
        String result = "";
        if (resultCode == Activity.RESULT_OK && data != null && data.getData() != null) {
            Uri uri = data.getData();
            // Persist the grant so the folder still works after the app restarts.
            // Without this the URI is only valid for this process lifetime, and the
            // player would have to re-pick the folder every single session.
            try {
                getActivity().getContentResolver().takePersistableUriPermission(uri,
                        Intent.FLAG_GRANT_READ_URI_PERMISSION
                                | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
            } catch (Exception ignored) {
            }
            result = uri.toString();
        }
        finishWith(result);
    }

    private void finishWith(String result) {
        UnityPlayer.UnitySendMessage(sCallbackObject, METHOD, result == null ? "" : result);
        try {
            getFragmentManager().beginTransaction().remove(this).commitAllowingStateLoss();
        } catch (Exception ignored) {
        }
    }

    // -------------------------------------------------------------- export

    /**
     * Copy srcPath into the chosen folder as displayName, replacing any existing
     * document with that name. Returns the document URI, or "" if anything failed.
     */
    public static String exportFile(String treeUriString, String srcPath,
                                    String displayName, String mime) {
        Activity activity = UnityPlayer.currentActivity;
        if (activity == null || treeUriString == null || srcPath == null
                || displayName == null || Build.VERSION.SDK_INT < 21) {
            return "";
        }
        InputStream in = null;
        OutputStream out = null;
        try {
            Uri tree = Uri.parse(treeUriString);
            String docId = DocumentsContract.getTreeDocumentId(tree);
            Uri dir = DocumentsContract.buildDocumentUriUsingTree(tree, docId);

            // Same name twice would otherwise create "log (1).csv", "log (2).csv"...
            // Delete first so re-exporting the same run always lands on one file.
            Uri existing = findChild(activity, tree, dir, displayName);
            if (existing != null) {
                try {
                    DocumentsContract.deleteDocument(activity.getContentResolver(), existing);
                } catch (Exception ignored) {
                }
            }
            Uri doc = DocumentsContract.createDocument(activity.getContentResolver(), dir,
                    mime == null || mime.length() == 0 ? "text/csv" : mime, displayName);
            if (doc == null) {
                return "";
            }
            in = new FileInputStream(srcPath);
            out = activity.getContentResolver().openOutputStream(doc);
            if (out == null) {
                return "";
            }
            byte[] buffer = new byte[65536];
            int read;
            while ((read = in.read(buffer)) > 0) {
                out.write(buffer, 0, read);
            }
            out.flush();
            return doc.toString();
        } catch (Exception e) {
            return "";
        } finally {
            try { if (in != null) in.close(); } catch (Exception ignored) { }
            try { if (out != null) out.close(); } catch (Exception ignored) { }
        }
    }

    private static Uri findChild(Activity activity, Uri tree, Uri dir, String displayName) {
        android.database.Cursor c = null;
        try {
            Uri children = DocumentsContract.buildChildDocumentsUriUsingTree(tree,
                    DocumentsContract.getDocumentId(dir));
            c = activity.getContentResolver().query(children, new String[] {
                    DocumentsContract.Document.COLUMN_DOCUMENT_ID,
                    DocumentsContract.Document.COLUMN_DISPLAY_NAME }, null, null, null);
            if (c == null) {
                return null;
            }
            while (c.moveToNext()) {
                if (displayName.equals(c.getString(1))) {
                    return DocumentsContract.buildDocumentUriUsingTree(tree, c.getString(0));
                }
            }
            return null;
        } catch (Exception e) {
            return null;
        } finally {
            try { if (c != null) c.close(); } catch (Exception ignored) { }
        }
    }

    /** Short readable name of the chosen folder, for the settings screen. */
    public static String folderLabel(String treeUriString) {
        if (treeUriString == null || treeUriString.length() == 0) {
            return "";
        }
        try {
            Uri tree = Uri.parse(treeUriString);
            String id = DocumentsContract.getTreeDocumentId(tree);
            if (id == null) {
                return treeUriString;
            }
            // Tree ids look like "primary:Download/logs" - the part after ':' is the path.
            int colon = id.indexOf(':');
            String path = colon >= 0 && colon + 1 < id.length() ? id.substring(colon + 1) : id;
            return path.length() == 0 ? "(root)" : path;
        } catch (Exception e) {
            return treeUriString;
        }
    }
}
